base/channels.jl

```diff
@@ -0,0 +1,181 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+abstract ChannelData{T}
+
+type ChannelDataSingle{T} <: ChannelData{T}
+    full::Bool          # Need a separate flag here to handle the case where "nothing" is a valid value in "data"
+    data::Nullable{T}
+end
+
+type ChannelDataMultiple{T} <: ChannelData{T}
+    data::Array{T,1}
+    szp1::Int               # current channel size plus one
+    sz_max::Int             # maximum size of channel
+    take_pos::Int           # read position
+    put_pos::Int            # write position
+end
+
+type ChannelAddr
+    pid::Int
+    cid::Int
+end
+
+abstract AbstractChannel{T}
+
+type Channel{T, Many} <: AbstractChannel{T}
+    cid::Int
+    store::ChannelData{T}
+    cond_take::Condition    # waiting for data to become available
+    cond_put::Condition     # waiting for a writeable slot
+
+    function Channel(Ty::Type, szp1::Int, sz_max::Int)
+        if sz_max==1
+            store = ChannelDataSingle{Ty}(false, Nullable{T}())
+        else
+            store = ChannelDataMultiple{Ty}(Array(Ty, szp1), szp1, sz_max, 1, 1)
+        end
+        new(get_next_channel_id(), store, Condition(), Condition())
+    end
+end
+
+let next_channel_id=1
+    global get_next_channel_id
+    function get_next_channel_id()
+        cid = next_channel_id
+        next_channel_id = next_channel_id+1
+        return cid
+    end
+end
+
+Channel() = Channel(1)
+Channel(sz::Int) = Channel(Any, sz)
+Channel(T::Type) = Channel(T, 1)
+function Channel(T::Type, sz::Int)
+    sz_max = sz == typemax(Int) ? typemax(Int) - 1 : sz
+    csz = sz > 32 ? 32 : sz
+    if sz_max == 1
+        Channel{T, false}(T, csz+1, sz_max)
+    else
+        Channel{T, true}(T, csz+1, sz_max)
+    end
+end
+
+map_all_channels = Dict{Int, AbstractChannel}()
+register_channel(c::Channel) = (map_all_channels[c.cid] = c; ChannelAddr(myid(), c.cid))
+create_and_register_channel(T::Type, sz::Int) = register_channel(Channel(T, sz))
+
+channel_at(pid::Int=myid(), T::Type=Any, sz::Int=1) = remotecall_fetch(pid, create_and_register_channel, T, sz)
+
+close_channel(cid::Int) = (delete!(map_all_channels, cid); nothing)
+close(addr::ChannelAddr) = remotecall_fetch(addr.pid, close_channel, addr.cid)
+
+# TODO Implement
+# put!(addr::ChannelAddr, v)
+# take!(addr::ChannelAddr)
+# fetch(addr::ChannelAddr)
+# isready(addr::ChannelAddr)
+# wait(addr::ChannelAddr)
+
+
+function put!{T}(c::Channel{T, true}, v::T)
+    store = c.store
+    d = store.take_pos - store.put_pos
+    if (d == 1) || (d == -(store.szp1-1))
```

Comment:
push! / pop! / shift! / unshift! should generally already take care of all of this